GoogleIoTCore.agent.nut

```diff
@@ -0,0 +1,1004 @@
+            "cache-control" : "no-cache"
+        };
+
+        local url = "https://cloudiot.googleapis.com/v1/";
+        url += format("projects/%s/locations/%s/registries/%s/devices", _projectId, _cloudRegion, _registryId);
+
+        local deviceDesc = {
+            "id": _deviceId,
+            "credentials": [
+                {
+                    "public_key": {
+                        // TODO: should we allow for another RSA format?
+                        "format": "RSA_X509_PEM",
+                        "key": publicKey
+                    }
+                }
+            ]
+        };
+        local data = http.jsonencode(deviceDesc);
+
+        local req = http.post(url, headers, data);
+
+        local sent = function (resp) {
+            if (resp.statuscode == 200) {
+                callback(0, null);
+            } else {
+                callback(resp.statuscode, resp.body);
+            }
+        }.bindenv(this);
+
+        req.sendasync(sent);
+    }
+
+    function _makeJwtToken(tokenReadyCb) {
+        local header = http.base64encode("{\"alg\":\"RS256\",\"typ\":\"JWT\"}");
+        local curTime = time();
+        local expTime = curTime + TOKEN_EXPIRATION_TIME;
+        local claimset = {
+            "aud"   : _projectId,
+            "exp"   : expTime,
+            "iat"   : curTime
+        };
+        local body = http.base64encode(http.jsonencode(claimset));
+
+        crypto.sign(crypto.RSASSA_PKCS1_SHA256, header + "." + body, _decodePem(_privateKey),
+            function(err, sig) {
+                if (err) {
+                    server.error(err);
+                    return;
+                }
+
+                local signature = http.base64encode(sig);
+
+                tokenReadyCb({
+                    "jwtToken" : (header + "." + body + "." + signature),
+                    "jwtExpiresAt" : expTime});
+
+            }.bindenv(this)
+        );
+    }
+
+    // Remove the armor, concatenate the lines, and base64 decode the text.
+    function _decodePem(str) {
+        local lines = split(str, "\n");
+        // We really ought to iterate over the array until we find a starting line,
+        // and then look for the matching ending line.
+        if ((lines[0] == "-----BEGIN PRIVATE KEY-----"
+                && lines[lines.len() - 1] == "-----END PRIVATE KEY-----") ||
+            (lines[0] == "-----BEGIN RSA PRIVATE KEY-----"
+                && lines[lines.len() - 1] == "-----END RSA PRIVATE KEY-----") ||
+            (lines[0] == "-----BEGIN PUBLIC KEY-----"
+                && lines[lines.len() - 1] == "-----END PUBLIC KEY-----"))
+        {
+            local all = lines.slice(1, lines.len() - 1).reduce(@(a, b) a + b);
+            return http.base64decode(all);
+        }
+        return null;
+    }
+
+    function _typeof() {
+        return "GoogleIoTCore.Client";
+    }
+}
+
+class GoogleIoTCore.AbstractTransport extends GoogleIoTCore {
```

Comment:
Why does it need to extend the `GoogleIoTCore` class?

Reply:
For example, we want to use `_log()` and `_logError()` methods from `GoogleIoTCore`

Reply:
And that's it? I prefer having a singleton log table in this cases. `Transport is a GoogleIoTCore` sounds weird`. But I'll leave it up to you.